### ISSUE TITLE
plugin for S3 compliant Cloud

### ISSUE THREAD

**Issue by dspeed2** [DataVault/datavault#405]
This maps to the following Jira issue

https://www.jira.is.ed.ac.uk/browse/RSS022-12

**Comment by dspeed2** [comment]
I've made a start on this.  I'm using the Amazon S3 API which I am not very clear on whether it is compatible with other S3 compliant storage (I suspect not) so we might have to rewrite when we decide on which cloud we are going to use.

I'm going to have to change things so that when we are using the cloud a new bucket is created for each vault.  I hope to keep all of this in the S3 plugin rather than the vault controller by doing something like if no bucket create at the beginning of a store action.  

I believe all buckets need to have cloud wide unique name so I'm planning on using the vault ID from the db with some sort of DV prefix (something like data_vault-5b38266f-dc6a-472c-b949-9ef905405302) this will also allow us to identify things in TSM / the cloud easily by matching the id in our db. 

I'm also going to make some changes as to how we store / retrieve stuff.  At the moment we use a random key to store and use the file path to retrieve.  For the new S3 plugin and probably the TSM plugin I'm going to change things so that the vault id is used as the storage key and we actually use that key in retrievals.  I'm hopefully going to manage to not change any of the existing plugins but we may want to upgrade those too if all of this works / makes sense.

This is all subject to my assumption that we are planning on having one bucket per vault and not one bucket per deposit if that is wrong we might want to use a combination of vault id and deposit key.

There are size limits on how much you can store in a bucket so we will want to keep an eye on that but at the moment I believe the limit is bigger than the max deposit we are planning on allowing.

I'm adding this to GitHub for a change to make it visible to everyone I'll copy it to Jira too.

**Comment by robintaylor** [comment]
Hi David, apologies if I am way out of touch here - Given that a vault may contain multiple deposits, should we not be using the deposit id as the key rather than the vault id?

**Comment by dspeed2** [comment]
I stupidly replied to this directly on the Github slack rather than here.  It also appears that I'd added most of the stuff I put on Slack to the ticket already in my previous comment.

It is still early days with this and I'm still working out an exact plan so the details in my previous comment will probably change. I will probably end up using the deposit key as the description in TSM and the retrieval key in the Cloud.  Then either the vault id with a DV prefix or a combination of vault id and deposit id for the Cloud bucket name.

Whatever happens I'll keep this updated so you can keep me right :-)

**Comment by stuartlewis** [comment]
I'm not sure if every new vault deserves a new bucket?  By default, AWS customers are allowed 100 buckets.  Any more than this requires a request to AWS.

 - https://docs.aws.amazon.com/AmazonS3/latest/dev/BucketRestrictions.html

By creating a new bucket for every vault, that quota of 100 could be very quickly used up.

Is there any reason not to use a single bucket, and then name the files <unique DV prefix>-<vault-id>-<deposit-id>.zip ? 

What added benefits are there by using multiple buckets?

**Comment by dspeed2** [comment]
Really? I thought you could create as many buckets as you wanted and you only got charged for any data you added.  I also read about bucket size limits which is why I was thinking of going even further and having a bucket per deposit.  A hard limit that would obviously change things.

I'll read your link and try and find the stuff I'd previously read then I'll update the ticket.

**Comment by dspeed2** [comment]
Ok the link Stuart sent is pretty explicit about the default of 100 buckets.  I believe that I must have misread the section in https://docs.aws.amazon.com/AmazonS3/latest/gsg/CreatingABucket.html where it says "you are not charged for creating a bucket"  as you can create lots of buckets.  

I also believe that I misread the section here https://docs.aws.amazon.com/AmazonS3/latest/dev/Introduction.html
"Store data in Buckets – Store an infinite amount of data in a bucket. Upload as many objects as you like into an Amazon S3 bucket. Each object can contain up to 5 TB of data. Each object is stored and retrieved using a unique developer-assigned key." to be something like a bucket not an object has a 5TB limit.

Ok I'm glad I added that comment here last week now!  This has been pretty embarrassing for me so far but at least people have pointed out the issues before I got to far with the code.

So without having had too much time to think about this we can just have one infinite datavault bucket.  A bucket per vault would have separated things nicely but it doesn't look like that will be an option.

**Comment by robintaylor** [comment]
Hi David, just some random thoughts - I'm assuming the S3 stuff is for JISC, do you know how they will handle the Amazon account side of things? One JISC account, or one account for each client? I'm thinking that the creation of a bucket could be tied to each client eg. datavault.randomuni.ac.uk. It could possibly be done as part of the setup process for a client. The DV app, or admin user, would then create an archive store with the name of the bucket and the credentials to access it. This would only need to be done one time.

**Comment by dspeed2** [comment]
Hey, the S3 stuff at the moment is the implementation of the 3rd copy of the deposits.  The reason  I am using S3 is just purely down to the actual cloud provider we are going to use hasn't been decided / purchased so I guess you should view this a prototype so we have 2 tape copies and a cloud copy available.  I was hoping that as whatever we pick will be S3 compliant we could just change the auth credentials and it would work but it is more likely the plugin will be rewritten or a new one will be written in addition to the S3 (which like you say might be useful for the JISC version) when the Cloud space is purchased.

After Stuart and your own comments yesterday I actually made a lot of progress yesterday.  I set up my own AWS account with a free year of limited S3 to include a dv test bucket, finished the first pass of the plugin and added it to the  database so that TSM and S3 archive stores are used with deposits.

I also updated the TSM and S3 plugins to use the deposit id rather than the vault id.

I still have some bugs to iron out but when I tried last night  on my local deploy I did manage to get all 3 copies stored.

There has been some discussion with JISC about the questions you asked I think it is still up in the air but I plan to move the bucket name etc for my S3 plugin into the properties file at some point so what you say seems plausible where the bucket is created for the client at setup and configured in the properties file.

**Comment by dspeed2** [comment]
This is just a copy of an update I've posted to lira about this

I decided to try and fix the issues in the previous comment by destroying my VM and rebuilding (including the TSM client) after a little bit of trouble I got that working again.

I spent most of yesterday working on this, the current state of play is that I have a S3 plugin that stores and retrieve objects in a single dv bucket using the deposit Id as a key.
I haven't written any unit tests yet for the plugin but I have manually tested things by doing the following

1) Initialise the app to have the TSM plugin as the only archive store in this state both store and retrieve work as expected (well two copies are stored and the first copy is retrieved.  The deposit id is used as the key.  I haven't tested the various other scenarios (missing copies / mismatched checksums) since the last time)
2) Initialise the app to have the S3 plugin as the only archive store in this state both store and retrieve work as expected (I can see the stored copy in the S3 bucket and the correct file is retrieved when requested)
3) Initialised the app to have both the TSM and S3 plugins as archive stores in this state store works as expected (two TSM copies and a S3 copy all keyed on deposit id) but retrieve gives a weird error at the RetrieveController stage (so doesn't even get to the worker) so far I haven't even been able to find any logs other than the pretty generic stack trace displayed in the browser.
I'll be working on fixing that and adding some unit tests next.

During this manual testing I've been using a bigger file (250ish MB) than I've previously used (just to try and simulate a 'realer' user experience the first deposit always reports an error (but the 3 copies are successfully stored) this appear to be something to do with the nice progress bar that displays when the bigger file is uploaded.

I'll try and get around to fixing that too.

**Comment by dspeed2** [comment]
Another copy from Jira concerning the fix for the retrieve deposit when multiple plugins are set up.

I've dug through the code I can see that the problem was due to the database config I'd set up.  I'd made both the TSM and the S3 plugin retrieve enabled, I can see from the comments in ArchiveStore that there should only be one plugin marked retrieve enabled ("// One ArchiveStore should be flagged as the preferred option for retrieving archives.").  I dunno yet if preferred means that the S3 plugin will be used if the TSM copies fail or if the S3 would be a manual retrieve in times of trouble.
Anyway that bit is fixed, I'll concentrate on adding tests and fixing the progress bar on the first deposit after a deploy.

**Comment by dspeed2** [comment]
Another copy from Jira

I've done a bit of experimenting with the progress bar it only happens for the first deposit after startup and it only happens if you choose a file, directories work fine.  Once the first 'failure' is done the next deposit works whether it is a file or dir or even the same file.  I did try adding a some sleeps to the code in case the issue was something not initialising quick enough but that didn't help.  I'm going to leave this for another time (and ticket)
I've run some manual tests on the app deployed with both TSM and S3 plugins included, apart from the above issue all deposits appear to work as does retrieve.  I've tested the retrieval if the first then first and second copies are missing and it did what we expected (retrieved the second copy or failed to retrieve leaving us to manually retrieve form the cloud)
I haven't added any thing other than logging to say that copies are missing or have the wrong checksums not attempt to repair or even alert admins is made.
